src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -691,72 +693,130 @@ putPaddingTo64 written
 -- Also note that the implementation reads values in little-endian byte order.
 -- If the file has been serialised in big-endian order, the mismatch will be
 -- detected by looking at the version indicator.
---
--- __NOTE__: Currently does not perform bounds checks. Malformed input can
--- trigger invalid memory access!
 fromSBS :: ShortByteString -> Either String (NumEntries, CompactIndex)
 fromSBS (SBS ba') = do
     let ba = ByteArray ba'
-    let len = sizeofByteArray ba
-    when (mod8 len /= 0) $ Left "Length is not multiple of 64 bit"
-    when (len < 36) $ Left "Too small for header and footer"
+    let len8 = sizeofByteArray ba
+    when (mod8 len8 /= 0) $ Left "Length is not multiple of 64 bit"
+    when (len8 < 36) $ Left "Doesn't contain header and footer"
 
+    -- check version
     let version = indexByteArray ba 0 :: Word32
     when (version == byteSwap32 indexVersion) $ Left "Non-matching endianness"
     when (version /= indexVersion) $ Left "Unsupported version"
 
-    let len64 = div8 len
-    let rfprec = fromIntegral (indexByteArray ba (len64 - 3) :: Word64)
-    let numPages = fromIntegral (indexByteArray ba (len64 - 2) :: Word64)
-    let numEntries = fromIntegral (indexByteArray ba (len64 - 1) :: Word64)
+    -- read footer
+    let len64 = div8 len8
+    let getPositive off64 = do
+          let w = indexByteArray ba off64 :: Word64
+          when (w > fromIntegral (maxBound :: Int)) $
+            Left "Size information is too large for Int"
+          return (fromIntegral w)
+
+    rfprec <- getPositive (len64 - 3)
+    numPages <- getPositive (len64 - 2)
+    numEntries <- getPositive (len64 - 1)
+
+    when (rfprec > snd rangeFinderPrecisionBounds) $
+      Left "Invalid range finder precision"
+    let numRanges = 2 ^ rfprec + 1
 
+    -- read vectors
     -- offsets in 32 bits
-    let offset1_32 = 1  -- after version indicator
-    let (offset2_32, ciPrimary) = getVec32 ba offset1_32 numPages
-    let (offset3_32, ciRangeFinder) = getVec32 ba offset2_32 (2 ^ rfprec + 1)
+    let off1_32 = 1  -- after version indicator
+    (!off2_32, ciPrimary) <- getVec32 "Primary array" ba off1_32 numPages
+    (!off3_32, ciRangeFinder) <- getVec32 "Range finder" ba off2_32 numRanges
     -- offsets in 64 bits
-    let offset3 = ceilDiv2 offset3_32
-    let (offset4, ciClashes) = getBitVec ba offset3 numPages
-    let (offset5, ciLargerThanPage) = getBitVec ba offset4 numPages
-    let ciTieBreaker = getTieBreaker ba offset5
+    let !off3 = ceilDiv2 off3_32
```

Comment:
```suggestion
    when (mod2 off3_32 /= 0) $ Left "Not enough padding after range finder component" 
    let !off3 = ceilDiv2 off3_32
```

Reply:
If the 32 bit offset after the tie breaker is not 64 bit aligned, that just means there are 32 bits of padding before the bit vectors start. This is what `ceilDiv` handles and it's totally expected.

Reply:
You're right